client/my-sites/stats/wordads-chart-tabs/index.jsx

```diff
@@ -0,0 +1,256 @@
+		return tooltipData;
+	}
+
+	onLegendClick = chartItem => {
+		const activeLegendCharts = this.state.activeLegendCharts;
+		const chartIndex = activeLegendCharts.indexOf( chartItem );
+		let gaEventAction;
+		if ( -1 === chartIndex ) {
+			activeLegendCharts.push( chartItem );
+			gaEventAction = ' on';
+		} else {
+			activeLegendCharts.splice( chartIndex );
+			gaEventAction = ' off';
+		}
+		this.props.recordGoogleEvent(
+			'WordAds Stats',
+			`Toggled Nested Chart ${ chartItem } ${ gaEventAction }`
+		);
+		this.setState( {
+			activeLegendCharts,
+		} );
+	};
+
+	getActiveTab( nextProps ) {
+		const props = nextProps || this.props;
+		return find( props.charts, { attr: props.chartTab } ) || props.charts[ 0 ];
+	}
+
+	getLoadedData() {
+		const { fullQueryData } = this.props;
+		return fullQueryData;
+	}
+
+	buildChartData() {
+		const data = this.getLoadedData();
+		if ( ! data ) {
+			return [];
+		}
+
+		const activeTab = this.props.chartTab;
+		const labelKey =
+			'label' +
+			this.props.period.period.charAt( 0 ).toUpperCase() +
+			this.props.period.period.slice( 1 );
+		return data.map( record => {
+			let recordClassName;
+			if ( record.classNames && record.classNames.length ) {
+				recordClassName = record.classNames.join( ' ' );
+			}
+
+			const className = classNames( recordClassName, {
+				'is-selected': record.period === this.props.queryDate,
+			} );
+
+			const item = {
+				label: record[ labelKey ],
+				value: record[ activeTab ],
+				data: record,
+				className,
+			};
+			item.tooltipData = this.buildTooltipData( item );
+
+			return item;
+		} );
+	}
+
+	render() {
+		const { query, fullQueryRequesting, siteId } = this.props;
+
+		const chartData = this.buildChartData();
+		const activeTab = this.getActiveTab();
+		const data = this.getLoadedData();
+
+		const activeTabLoading = fullQueryRequesting && ! ( data && data.length );
+		const classes = [
+			'stats-module',
+			'is-chart-tabs',
+			{
+				'is-loading': activeTabLoading,
+			},
+		];
+
+		return (
+			<div>
+				{ siteId && <QuerySiteStats statType="statsAds" siteId={ siteId } query={ query } /> }
+
+				<Card className={ classNames( ...classes ) }>
+					<Legend
+						tabs={ this.props.charts }
+						activeTab={ activeTab }
+						activeCharts={ this.state.activeLegendCharts }
+						clickHandler={ this.onLegendClick }
+					/>
+					<StatsModulePlaceholder className="is-chart" isLoading={ activeTabLoading } />
+					<ElementChart
+						loading={ activeTabLoading }
+						data={ chartData }
+						barClick={ this.props.barClick }
+					/>
+					<StatTabs
+						data={ data }
+						tabs={ this.props.charts }
+						switchTab={ this.props.switchTab }
+						selectedTab={ this.props.chartTab }
+						activeIndex={ this.props.queryDate }
+						activeKey="period"
+					/>
+				</Card>
+			</div>
+		);
+	}
+}
+
+const connectComponent = connect(
+	( state, { moment, period: periodObject, queryDate } ) => {
+		const siteId = getSelectedSiteId( state );
```

Comment:
this is a lot of work to be doing in a connect method. This runs every time the react store changes and this component is connected. Could we memoize some of the results so we don't have to recalculate it all every time?

Reply:
Took a little while to wrap my head around this, but I think I've got it working?
@blowery thanks for the link on derived state above - the memoization section there helped a lot in figuring this part out.

I threw a `console.log()` into the buildQuery() function while testing and was able to see just how many fewer times this was running, which was pretty neat 😃